lib/elixir/lib/calendar/date.ex

```diff
@@ -757,6 +769,75 @@ defmodule Date do
     end
   end
 
+  @doc """
+  Shifts a date by given units. Raises an `ArgumentError` if the
+  given date is not a valid Calendar.ISO date.
+
+  Available shift units are: `:year, :month, :week, :day` and the shift
+  is always applied in that order.
+
+  `Date.shift/2` defines a week as 7 days and a year as 12 months.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - if the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Date.shift(~D[2016-01-03], month: 2)
+      ~D[2016-03-03]
+      iex> Date.shift(~D[2016-02-29], month: 1)
+      ~D[2016-03-29]
+      iex> Date.shift(~D[2016-01-31], month: 1)
+      ~D[2016-02-29]
+      iex> Date.shift(~D[2016-01-31], year: 4, day: 1)
+      ~D[2020-02-01]
+
+  """
+  # @doc since: "1.5.0"
+  @spec shift(Calendar.date(), [shift_unit()]) :: t
+  def shift(%{calendar: Calendar.ISO} = date, shift_units) do
+    shift_units = Keyword.validate!(shift_units, year: 0, month: 0, week: 0, day: 0)
+
+    ordered_units = [
+      year: shift_units[:year],
+      month: shift_units[:month],
+      week: shift_units[:week],
+      day: shift_units[:day]
+    ]
```

Comment:
We need to discuss if we should indeed order the fields. That's because 1year + 1month can be different than 1month + 1year. If we order the fields, then we can express only one of them in a single computation. In order words, I would traverse the given shift_units directly without reordering.

Reply:
That makes sense, i was torn between ordering or not and went for consistency over flexibility. The more i think about it and reading your comment, I agree it should be up to the caller.

Reply:
After thinking about it more and discussing with @wojtekmach, I have actually changed my mind. The issue is that Elixir does not actually guarantee an order for operations in `Keyword.put/3`. Therefore, if you `Keyword.put/3` a month into the options, you won't actually know the result without inspecting it. So I think your original approach is better.

Reply:
💡 i am still torn to be honest. How about using the ordered version by default, but allow the caller to use the given order if they want to? Not sure if that's worth supporting tbh